pkg/autoupdate/tuf/library_manager.go

```diff
@@ -320,53 +246,246 @@ func (ulm *updateLibraryManager) tidyUpdateLibrary(binary autoupdatableBinary, c
 
 	const numberOfVersionsToKeep = 3
 
-	rawVersionsInLibrary, err := filepath.Glob(filepath.Join(ulm.updatesDirectory(binary), "*"))
+	versionsInLibrary, invalidVersionsInLibrary, err := ulm.sortedVersionsInLibrary(binary)
 	if err != nil {
-		level.Debug(ulm.logger).Log("msg", "could not glob for updates to tidy updates library", "err", err)
+		level.Debug(ulm.logger).Log("msg", "could not get versions in library to tidy update library", "err", err)
 		return
 	}
 
-	versionsInLibrary := make([]*semver.Version, 0)
-	for _, rawVersion := range rawVersionsInLibrary {
-		v, err := semver.NewVersion(filepath.Base(rawVersion))
-		if err != nil {
-			level.Debug(ulm.logger).Log("msg", "updates library contains invalid semver", "err", err, "library_path", rawVersion)
-			ulm.removeUpdate(binary, filepath.Base(rawVersion))
-			continue
-		}
-
-		versionsInLibrary = append(versionsInLibrary, v)
+	for _, invalidVersion := range invalidVersionsInLibrary {
+		level.Debug(ulm.logger).Log("msg", "updates library contains invalid version", "err", err, "library_path", invalidVersion)
+		ulm.removeUpdate(binary, invalidVersion)
 	}
 
 	if len(versionsInLibrary) <= numberOfVersionsToKeep {
 		return
 	}
 
-	// Sort the versions (ascending order)
-	sort.Sort(semver.Collection(versionsInLibrary))
-
 	// Loop through, looking at the most recent versions first, and remove all once we hit nonCurrentlyRunningVersionsKept valid executables
 	nonCurrentlyRunningVersionsKept := 0
 	for i := len(versionsInLibrary) - 1; i >= 0; i -= 1 {
 		// Always keep the current running executable
-		if versionsInLibrary[i].Original() == currentRunningVersion {
+		if versionsInLibrary[i] == currentRunningVersion {
 			continue
 		}
 
 		// If we've already hit the number of versions to keep, then start to remove the older ones.
 		// We want to keep numberOfVersionsToKeep total, saving a spot for the currently running version.
 		if nonCurrentlyRunningVersionsKept >= numberOfVersionsToKeep-1 {
-			ulm.removeUpdate(binary, versionsInLibrary[i].Original())
+			ulm.removeUpdate(binary, versionsInLibrary[i])
+			continue
+		}
+
+		nonCurrentlyRunningVersionsKept += 1
+	}
+}
+
+// sortedVersionsInLibrary looks through the update library for the given binary to validate and sort all
+// available versions. It returns a sorted list of the valid versions, a list of invalid versions, and
+// an error only when unable to glob for versions.
+func (ulm *updateLibraryManager) sortedVersionsInLibrary(binary autoupdatableBinary) ([]string, []string, error) {
+	rawVersionsInLibrary, err := filepath.Glob(filepath.Join(ulm.updatesDirectory(binary), "*"))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not glob for updates in library: %w", err)
+	}
+
+	versionsInLibrary := make([]*semver.Version, 0)
+	invalidVersions := make([]string, 0)
+	for _, rawVersionWithPath := range rawVersionsInLibrary {
+		rawVersion := filepath.Base(rawVersionWithPath)
+		v, err := semver.NewVersion(rawVersion)
```

Comment:
I haven't tested this, but how does it handle the git versions we use? `v1.0.7-16-g6e6704e1dc33`

Reply:
It parses them and sorts them correctly, I think. It will say that `1.0.7-16-abcdabcd` is more recent than `1.0.7-15-defgdefg`. It also says that `1.0.7-16-g6e6704e1dc33` is _not_ more recent than `1.0.7`, which looks like it's expected and adheres to the semver spec, but is maybe not what we expect?

Reply:
(There's a test `Test_sortedVersionsInLibrary` that covers at least some of these test cases -- I'm happy to expand it too.)